pkg/util/pki/csr.go

```diff
@@ -278,148 +302,208 @@ func buildKeyUsagesExtensionsForCertificate(crt *v1.Certificate) ([]pkix.Extensi
+		return nil, err
 	}
 
-	certDuration := apiutil.DefaultCertDuration(crt.Spec.Duration)
-
-	pubKeyAlgo, _, err := SignatureAlgorithm(crt)
+	certDuration := apiutil.DefaultCertDuration(cr.Spec.Duration)
+	keyUsage, extKeyUsage, err := KeyUsagesForCertificateOrCertificateRequest(cr.Spec.Usages, cr.Spec.IsCA)
 	if err != nil {
 		return nil, err
 	}
 
-	cert := &x509.Certificate{
-		// Version must be 2 according to RFC5280.
-		// A version value of 2 confusingly means version 3.
-		// This value isn't used by Go at the time of writing.
-		// https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.1
-		Version:               2,
-		BasicConstraintsValid: true,
-		SerialNumber:          serialNumber,
-		PublicKeyAlgorithm:    pubKeyAlgo,
-		IsCA:                  crt.Spec.IsCA,
-		NotBefore:             time.Now(),
-		NotAfter:              time.Now().Add(certDuration),
-		// see http://golang.org/pkg/crypto/x509/#KeyUsage
-		KeyUsage:       keyUsages,
-		ExtKeyUsage:    extKeyUsages,
-		DNSNames:       dnsNames,
-		IPAddresses:    ipAddresses,
-		URIs:           uris,
-		EmailAddresses: crt.Spec.EmailAddresses,
-	}
+	return UpdateTemplateFromOptions(cert, certDuration, cr.Spec.IsCA, nil, keyUsage, extKeyUsage), nil
+}
 
-	if isLiteralCertificateSubjectEnabled() && len(crt.Spec.LiteralSubject) > 0 {
-		rawSubject, err := ParseSubjectStringToRawDERBytes(crt.Spec.LiteralSubject)
-		if err != nil {
-			return nil, err
-		}
+func UpdateTemplateFromOptions(cert *x509.Certificate, duration time.Duration, isCA bool, maxPathLen *int, keyUsage x509.KeyUsage, extKeyUsage []x509.ExtKeyUsage) *x509.Certificate {
+	// Override certificate template values with values that are expected but
+	// not present in the CSR that generated the certificate template.
 
-		cert.RawSubject = rawSubject
+	cert.NotBefore = time.Now()
+	cert.NotAfter = cert.NotBefore.Add(duration)
+
+	// We expect that if the CSR contains a BasicConstraints extension, it
+	// matches the IsCA and MaxPathLen values that we set here.
+	// This should have been enforced by our validation webhook.
+
+	cert.BasicConstraintsValid = true
+	cert.IsCA = isCA
+	if maxPathLen != nil {
+		cert.MaxPathLen = *maxPathLen
+		cert.MaxPathLenZero = *maxPathLen == 0
 	} else {
-		cert.Subject = pkix.Name{
-			Country:            subject.Countries,
-			Organization:       organization,
-			OrganizationalUnit: subject.OrganizationalUnits,
-			Locality:           subject.Localities,
-			Province:           subject.Provinces,
-			StreetAddress:      subject.StreetAddresses,
-			PostalCode:         subject.PostalCodes,
-			SerialNumber:       subject.SerialNumber,
-			CommonName:         commonName,
+		cert.MaxPathLen = 0
+		cert.MaxPathLenZero = false
+	}
+
+	// We expect that if the CSR contains a KeyUsage/ ExtendedKeyUsage
+	// extension, these KeyUsages match the usages that we set here. To
+	// be safe, we append the usages instead of replacing them.
```

Comment:
Maybe we should replace the KeyUsages and ExtendedKeyUsages here?

Reply:
I think given the function is currently called "override" I'd probably expect the arguments I passed to be used, replacing anything which was previously set.

Reply:
I updated the code and added modifier functions, these functions allow you to choose between an "override" keyusages and and "add" keyusages function.